astropy/table/table.py

```diff
@@ -3283,18 +3283,16 @@ def _encode_mixins(tbl):
         for name, column in tbl.columns.items():
             if isinstance(column, MaskedColumn) and np.any(column.mask):
                 if column.dtype.kind in ['i', 'u']:
-                    out[name] = column.astype(float).filled(np.nan)
-                    warnings.warn(
-                        "converted column '{}' from integer to float".format(
-                            name), TableReplaceWarning, stacklevel=3)
+                    out[name] = Series(column, dtype='Int64')
```

Comment:
The dtype should match the input dtype, e.g. `np.uint8` should go to the same (`pd.UInt8Dtype`) and so on.